File: day04/main.py
def main():
    with open("input.txt", "r") as f:
        data = f.readlines()

    data = [l.strip() for l in data]

    result = first(data)
    print(f"Result 1: {result}")

    result = second(data)
    print(f"Result 2: {result}")


def first(lines: list[str]):
    count = 0
    for i, line in enumerate(lines):
        for j, char in enumerate(line):
            if char == "X":
                words = diagonal_words(lines, i, j)
                count += len(words)
    return count


def diagonal_words(lines: [str], i: int, j: int) -> list[str]:
    word_size = 4
    found_words = [lines[i][j] for _ in range(8)]
    for d in range(1, word_size):
        # right
        next_col = j + d
        if next_col < len(lines[i]):
            found_words[0] += lines[i][next_col]

        # down
        new_row = i + d
        if new_row < len(lines):
            found_words[1] += lines[new_row][j]

        # left
        next_col = j - d
        if next_col >= 0:
            found_words[2] += lines[i][next_col]

        # up
        new_row = i - d
        if new_row >= 0:
            found_words[3] += lines[new_row][j]

        # down_right
        next_row = i + d
        next_col = j + d
        if next_row < len(lines) and next_col < len(lines[next_row]):
            found_words[4] += lines[next_row][next_col]

        # up right
        next_row = i - d
        next_col = j + d
        if next_row >= 0 and next_col < len(lines[next_row]):
            found_words[5] += lines[next_row][next_col]

        # down left
        next_row = i + d
        next_col = j - d
        if next_row < len(lines) and next_col >= 0:
            found_words[6] += lines[next_row][next_col]

        # up left
        next_row = i - d
        next_col = j - d
        if next_row >= 0 and next_col >= 0:
            found_words[7] += lines[next_row][next_col]

    words = [f for f in found_words if f == "XMAS"]
    return words


def second(lines: list[str]):
    count = 0
    for i, line in enumerate(lines):
        for j, char in enumerate(line):
            if char == "A" and 0 < i < len(lines) - 1 and 0 < j < len(line) - 1:
                if is_diagonal_cross(lines, i, j):
                    count += 1

    return count


def is_diagonal_cross(lines: [str], i: int, j: int) -> bool:
    crosses = [
        [
            ['M', ".", "S"],
            [".", "A", "."],
            ['M', ".", "S"],
        ],
        [
            ['M', ".", "M"],
            [".", "A", "."],
            ['S', ".", "S"],
        ],
        [
            ['S', ".", "M"],
            [".", "A", "."],
            ['S', ".", "M"],
        ],
        [
            ['S', ".", "S"],
            [".", "A", "."],
            ['M', ".", "M"],
        ],
    ]
    for cross in crosses:
        if (
                lines[i - 1][j - 1] == cross[0][0]
                and lines[i - 1][j + 1] == cross[0][2]
                and lines[i + 1][j - 1] == cross[2][0]
                and lines[i + 1][j + 1] == cross[2][2]
        ):
            return True
    return False


if __name__ == '__main__':
    main()

File: day04/main.zig
const std = @import("std");

pub fn main() !void {
    // according to documentation https://ziglang.org/documentation/0.13.0/#Choosing-an-Allocator
    // change the allocator to this makes more sense
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const buffer = try std.fs.cwd().readFileAlloc(allocator, "input.txt", 1024 * 1024);
    defer allocator.free(buffer);

    var lines = std.ArrayList([]const u8).init(allocator);
    var iter = std.mem.split(u8, buffer, "\n");
    while (iter.next()) |line| {
        if (line.len == 0) {
            continue;
        }
        try lines.append(line);
    }

    var result = try first(allocator, lines.items);
    std.debug.print("Result 1: {d}\n", .{result});

    result = try second(allocator, lines.items);
    std.debug.print("Result 2: {d}\n", .{result});
}

fn first(_: std.mem.Allocator, lines: [][]const u8) !i32 {
    var total: i32 = 0;
    for (0..lines.len) |row| {
        for (0..lines[row].len) |col| {
            std.debug.assert(lines.len == lines[row].len);
            if (lines[row][col] == 'X') {
                const current = Pos{ .row = @intCast(row), .col = @intCast(col) };

                inline for (@typeInfo(Direction).Enum.fields) |field| {
                    if (searchWordInDirection(lines, current, lines.len, @enumFromInt(field.value))) {
                        total += 1;
                    }
                }
            }
        }
    }

    return total;
}

fn searchWordInDirection(lines: [][]const u8, start: Pos, size: usize, direction: Direction) bool {
    var word = [4]u8{ 'X', ' ', ' ', ' ' };
    for (1..4) |multiplier| {
        const new_pos = start.move(direction, @intCast(multiplier), size) orelse {
            return false;
        };
        const new_char = lines[@intCast(new_pos.row)][@intCast(new_pos.col)];
        word[multiplier] = new_char;

        switch (multiplier) {
            1 => {
                if (!std.mem.eql(u8, word[0..], "XM  ")) {
                    return false;
                }
            },
            2 => {
                if (!std.mem.eql(u8, word[0..], "XMA ")) {
                    return false;
                }
            },
            3 => {
                if (!std.mem.eql(u8, word[0..], "XMAS")) {
                    return false;
                }
            },
            else => unreachable,
        }
    }
    return true;
}

fn second(_: std.mem.Allocator, lines: [][]const u8) !i32 {
    var total: i32 = 0;
    for (1..lines.len - 1) |row| {
        for (1..lines[row].len - 1) |col| {
            std.debug.assert(lines.len == lines[row].len);
            if (lines[row][col] == 'A') {
                const current = Pos{ .row = @intCast(row), .col = @intCast(col) };
                if (isDiagonalCross(lines, current, lines.len)) {
                    total += 1;
                }
            }
        }
    }

    return total;
}

fn isDiagonalCross(lines: [][]const u8, start: Pos, size: usize) bool {
    const start_cross = Pos{ .row = 1, .col = 1 };
    for (crosses) |cross| {
        const directions = [_]Direction{ .left_up, .right_up, .right_down, .left_down };

        var is_cross = true;
        for (directions) |direction| {
            const lines_pos = start.move(direction, 1, size) orelse {
                return false;
            };
            const cross_pos = start_cross.move(direction, 1, 3) orelse {
                return false;
            };
            if (lines[@intCast(lines_pos.row)][@intCast(lines_pos.col)] != cross[@intCast(cross_pos.row)][@intCast(cross_pos.col)]) {
                is_cross = false;
            }
        }
        if (is_cross) {
            return true;
        }
    }

    return false;
}

const crosses = [4][3][3]u8{
    [3][3]u8{
        [3]u8{ 'M', '.', 'S' },
        [3]u8{ '.', 'A', '.' },
        [3]u8{ 'M', '.', 'S' },
    },
    [3][3]u8{
        [3]u8{ 'M', '.', 'M' },
        [3]u8{ '.', 'A', '.' },
        [3]u8{ 'S', '.', 'S' },
    },
    [3][3]u8{
        [3]u8{ 'S', '.', 'M' },
        [3]u8{ '.', 'A', '.' },
        [3]u8{ 'S', '.', 'M' },
    },
    [3][3]u8{
        [3]u8{ 'S', '.', 'S' },
        [3]u8{ '.', 'A', '.' },
        [3]u8{ 'M', '.', 'M' },
    },
};

const Pos = struct {
    row: i32,
    col: i32,

    pub fn move(self: Pos, direction: Direction, multiplier: i32, size: usize) ?Pos {
        switch (direction) {
            .up => {
                if (self.row - 1 * multiplier < 0) {
                    return null;
                }
                return Pos{ .row = self.row - 1 * multiplier, .col = self.col };
            },
            .right_up => {
                if (self.row - 1 * multiplier < 0) {
                    return null;
                }
                if (self.col + 1 * multiplier > size - 1) {
                    return null;
                }
                return Pos{ .row = self.row - 1 * multiplier, .col = self.col + 1 * multiplier };
            },
            .right => {
                if (self.col + 1 * multiplier > size - 1) {
                    return null;
                }
                return Pos{ .row = self.row, .col = self.col + 1 * multiplier };
            },
            .right_down => {
                if (self.row + 1 * multiplier > size - 1) {
                    return null;
                }
                if (self.col + 1 * multiplier > size - 1) {
                    return null;
                }
                return Pos{ .row = self.row + 1 * multiplier, .col = self.col + 1 * multiplier };
            },
            .down => {
                if (self.row + 1 * multiplier > size - 1) {
                    return null;
                }
                return Pos{ .row = self.row + 1 * multiplier, .col = self.col };
            },
            .left_down => {
                if (self.row + 1 * multiplier > size - 1) {
                    return null;
                }
                if (self.col - 1 * multiplier < 0) {
                    return null;
                }
                return Pos{ .row = self.row + 1 * multiplier, .col = self.col - 1 * multiplier };
            },
            .left => {
                if (self.col - 1 * multiplier < 0) {
                    return null;
                }
                return Pos{ .row = self.row, .col = self.col - 1 * multiplier };
            },
            .left_up => {
                if (self.row - 1 * multiplier < 0) {
                    return null;
                }
                if (self.col - 1 * multiplier < 0) {
                    return null;
                }
                return Pos{ .row = self.row - 1 * multiplier, .col = self.col - 1 * multiplier };
            },
        }
    }
};

const Direction = enum {
    up,
    right_up,
    right,
    right_down,
    down,
    left_down,
    left,
    left_up,
};
